cardano-node/src/Cardano/Node/Shutdown.hs

```diff
@@ -0,0 +1,191 @@
+--
+-- The file descriptor wrapped in a 'ShutdownListener' designates the
+-- receiving end of the shutdown signalling communication channel.
+-- The opposite end might be either internal or external to the node process.
+withShutdownHandler :: ShutdownFDs -> Trace IO Text -> IO () -> IO ()
+withShutdownHandler sfds trace action
+  | Just (ShutdownListener fd) <- sfdsListener sfds =
+      Async.race_ (wrapUninterruptableIO $ waitForEOF fd) action
+  | otherwise = action
+  where
+    waitForEOF :: Fd -> IO ()
+    waitForEOF (Fd fd) = do
+      hnd <- IO.fdToHandle fd
+      r   <- try $ IO.hGetChar hnd
+      case r of
+        Left e
+          | IO.isEOFError e -> traceWith tracer "received shutdown request"
+          | otherwise       -> throwIO e
+
+        Right _  ->
+          throwIO $ IO.userError "--shutdown-ipc FD does not expect input"
+
+    sfdsListener :: ShutdownFDs -> Maybe ShutdownListener
+    sfdsListener = \case
+      ExternalShutdown r -> Just r
+      InternalShutdown r _w -> Just r
+      _ -> Nothing
+
+    tracer :: Tracer IO Text
+    tracer = trTransformer TextualRepresentation MaximalVerbosity
+                           (severityNotice trace)
+
+-- | Windows blocking file IO calls like 'hGetChar' are not interruptable by
+-- asynchronous exceptions, as used by async 'cancel' (as of base-4.12).
+--
+-- This wrapper works around that problem by running the blocking IO in a
+-- separate thread. If the parent thread receives an async cancel then it
+-- will return. Note however that in this circumstance the child thread may
+-- continue and remain blocked, leading to a leak of the thread. As such this
+-- is only reasonable to use a fixed number of times for the whole process.
+--
+wrapUninterruptableIO :: IO a -> IO a
+wrapUninterruptableIO action = async action >>= wait
+
+-- | If 'ShutdownFDs' supports internal shutdown requests,
+-- return its shutdown doorbell.
+getShutdownDoorbell :: ShutdownFDs -> Maybe ShutdownDoorbell
+getShutdownDoorbell (InternalShutdown _l doorbell) = Just doorbell
+getShutdownDoorbell _ = Nothing
+
+-- | Given the 'ShutdownDoorbell' component of 'ShutdownFDs',
+--   and an explanation of the reason, request a graceful shutdown.
+triggerShutdown :: ShutdownDoorbell -> Trace IO Text -> Text -> IO ()
+triggerShutdown (ShutdownDoorbell (Fd shutFd)) trace reason = do
+  traceWith (trTransformer TextualRepresentation MaximalVerbosity $
+              severityNotice trace)
+    ("Ringing the node shutdown doorbell:  " <> reason)
+  IO.hClose =<< IO.fdToHandle shutFd
+
+-- | We provide an optional cross-platform method to politely request shut down.
+--
+-- For the duration of 'action', we gracefully handle shutdown requests,
+-- external or internal, as requested by configuration in 'NodeProtocolMode',
+-- while allocating corresponding 'ShutdownFDs', and providing them to the 'action'.
+withShutdownHandling
+  :: NodeProtocolMode
+  -> Trace IO Text
+  -> (ShutdownFDs -> IO ())
+  -> IO ()
```

Comment:
Note how this is specifically not merged with `withShutdownHandler`:

- `withShutdownHandler` remains the mechanism
- `withShutdownHandling` is the policy bit around the mechanism